### PR TITLE
Theia plugin metadata

### PR DESCRIPTION
**Pull request by akurinnoy**
this PR implements
- retrieving metadata from plugins and using the metadata for plugin activation and deactivation;
- pluggability, which allows extending the plugin system to run different types of plugins.

### CODE REVIEW

File: packages/plugin/src/common/plugin-protocol.ts [hunk 1]
```diff
@@ -9,14 +9,125 @@
...
+import { RPCProtocol } from '../api/rpc-protocol';
+import { Disposable } from '../plugin/types-impl';
 
 export const hostedServicePath = '/services/hostedPlugin';
 
-export interface Plugin {
+/**
+ * Plugin engine (API) type, i.e. 'theia', 'vscode', etc.
+ */
+export type pluginEngine = string;
```
Comment (benoitf) [inline packages/plugin/src/common/plugin-protocol.ts]: should the type should start with an uppercase ?

File: packages/plugin/src/common/plugin-protocol.ts [hunk 2]
```diff
@@ -9,14 +9,125 @@
...
+export interface PluginPackage {
+    name: string;
+    publisher: string;
+    version: string;
+    engines: {
+        [type in pluginEngine]: string;
+    };
+    theiaPlugin?: {
+        frontend?: string;
+        frontendModuleName?: string;
```
Comment (benoitf) [inline packages/plugin/src/common/plugin-protocol.ts]: should we allow frontendModuleName from package.json or it's just inside the PluginLifecycle that we can get it?

File: packages/plugin/src/common/plugin-protocol.ts [hunk 3]
```diff
@@ -9,14 +9,125 @@
...
+        backend?: string;
+    };
+    main?: string;
+    displayName: string;
+    description: string;
+    contributes: {};
```
Comment (benoitf) [inline packages/plugin/src/common/plugin-protocol.ts:35]: what is contributes ?

File: packages/plugin/src/node/scanners/scanner-theia.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        };
+    }
+
+    getLifecycle(plugin: PluginPackage): PluginLifecycle {
+        return {
+            startMethod: 'start',
+            stopMethod: 'stop',
+            frontendModuleName: plugin.theiaPlugin!.frontendModuleName,
+
+            backendInitPath: 'backend-init-theia.js'
```
Comment (benoitf) [inline packages/plugin/src/node/scanners/scanner-theia.ts]: is there a reason why it's a .js file and not an interface with inversify binding ?

like BackendInit with theia and vscode implementation ?

so we need to bind a scanner + a init provider for each plugin type ?

File: packages/plugin/src/common/plugin-protocol.ts (hunk 1 above)
Comment (akurinnoy) [reply packages/plugin/src/common/plugin-protocol.ts]: yes, you're right, I'll fix it

File: packages/plugin/src/common/plugin-protocol.ts (hunk 2 above)
Comment (akurinnoy) [reply packages/plugin/src/common/plugin-protocol.ts]: I've added that field accidentally and forgot to remove it.
We do not take into account frontendModuleName from package.json for now.

File: packages/plugin/src/common/plugin-protocol.ts (hunk 3 above)
Comment (akurinnoy) [reply packages/plugin/src/common/plugin-protocol.ts:35]: It's added for future use. It is an object which describes contribution points of a plugin.

File: packages/plugin/src/node/scanners/scanner-theia.ts (hunk 4 above)
Comment (akurinnoy) [reply packages/plugin/src/node/scanners/scanner-theia.ts]: There is no particular reason for that, so, I'll revamp this part in the proposed way.

**Comment by benoitf** [comment]
FYI it seems there is a conflicting file on the PR as well

File: packages/plugin/src/node/scanners/scanner-theia.ts (hunk 4 above)
Comment (akurinnoy) [reply packages/plugin/src/node/scanners/scanner-theia.ts]: @benoitf 
I've started implementing your proposal but then remember that this script has to be runned on plugin host (webworker or node). So code cannot be passed through json-rpc connection.

Comment (benoitf) [reply packages/plugin/src/node/scanners/scanner-theia.ts]: @akurinnoy on backend it's still sent to webworker ?

Comment (akurinnoy) [reply packages/plugin/src/node/scanners/scanner-theia.ts]: not webworker but separate nodejs process

Comment (benoitf) [reply packages/plugin/src/node/scanners/scanner-theia.ts]: ok

**Review by evidolob — APPROVED** [review]

**Review by benoitf — APPROVED** [review]
if you can squash on merge, thx

**Pull request merged**